### DISCUSSION TITLE
How to resolveComponent when route query params change

### DISCUSSION THREAD

**Discussion by milelazar** [nuxt/nuxt#23358] [Questions]
I have a [...slug].vue catch all page with dynamic component that loads other templates/components based on API response.

Whenever url changes, page reruns and component is loaded using resolveComponent. The problem is when only query parameters change. In that case, page is not triggered (eg. you are on /products/ page and you click on a link with /products/?query=something url)

I have tried watching route with watchEffect and resolveComponent whenever url including query parameter change. And it kinda works but with "resolveComponent can only be used in render() or setup()" warning.

Is there a way to "rerun" the page when query changes without watch/watchEffect, or at least to get rid of the warning?

Here is a minimal reproduction: https://stackblitz.com/edit/github-pohmqb?file=pages%2F%5B...slug%5D.vue
When you click around on navigation links, correct component is loaded but with console warnings.

route.params.slug is sometimes returned as an Array, sometimes as a String.
resoloveComponent is also sometimes returned as AsyncComponentWrapper object or as a String when there is a query param.
Not sure what is happening.

**Comment by AloisSeckar** [comment]
AFAIK you can't use variable inside `resolveComponent` in Nuxt ([check the docs](https://nuxt.com/docs/guide/directory-structure/components#dynamic-components)). So maybe what you need is a switch with different `resolveComponent` calls.

This is how I ended up doing it in my project:

```
switch (podformular?.toLowerCase()) {
    case 'mp':
      return resolveComponent('Zadost50PodformularMP')
    case 'mpdavky':
      return resolveComponent('Zadost50PodformularMPDavky')

    // more case labels for respective components

    default:
      // handle invalid value
  }
```

**Comment by milelazar** [comment]
You can use variable inside resolveComponent, but only if component is registered globally. Works fine.

I have bunch of components so switch would get long and messy but I managed to solve the problem by ditching resolveComponent and using simple computed property. Not sure why I haven't tried that simplest solution before but it works without issues.
Al least with minimal reproduction https://stackblitz.com/edit/github-pohmqb-9edk9f?file=pages%2F%5B...slug%5D.vue

Will see if there would be any issues on real project.
